### PR TITLE
Use the unix-errno package for detecting and reporting errors.

### PR DESCRIPTION
**Pull request by yallop**

### CODE REVIEW

**Comment by yallop** [comment]
Here's an example.

Before:

```ocaml
# Unix_unistd.unlink "/does-not-exist";;
Exception: Unix.Unix_error (Unix.ENOENT, "unlink", "/does-not-exist").
```

After:

```ocaml
# Unix_unistd.unlink "/does-not-exist";;
Exception:
Errno.Error
 {Errno.errno = [Errno.ENOENT]; call = "unlink"; label = "/does-not-exist"}.
```

File: lib/ctypes/unix_unistd.ml [hunk 1]
```diff
@@ -68,115 +68,125 @@ let lseek =
...
 let unlink =
-  let c = local ~check_errno:true (unix_unistd_unlink_ptr ())
+  let c = local (unix_unistd_unlink_ptr ())
     PosixTypes.(string @-> returning int)
   in
   fun pathname ->
-    try ignore (c pathname)
-    with Unix.Unix_error(e,_,_) -> raise (Unix.Unix_error (e,"unlink",pathname))
+    Errno_unix.raise_on_errno ~call:"unlink" ~label:pathname begin fun () ->
+    ignore (c pathname)
```
Comment (dsheets) [inline lib/ctypes/unix_unistd.ml]: indent, please

File: lib/ctypes/unix_unistd.ml [hunk 2]
```diff
@@ -68,115 +68,125 @@ let lseek =
...
 let rmdir =
-  let c = local ~check_errno:true (unix_unistd_rmdir_ptr ())
+  let c = local (unix_unistd_rmdir_ptr ())
     PosixTypes.(string @-> returning int)
   in
   fun pathname ->
-    try ignore (c pathname)
-    with Unix.Unix_error(e,_,_) -> raise (Unix.Unix_error (e,"rmdir",pathname))
+    Errno_unix.raise_on_errno ~call:"rmdir" ~label:pathname begin fun () ->
+    ignore (c pathname)
```
Comment (dsheets) [inline lib/ctypes/unix_unistd.ml]: and here

File: lib/ctypes/unix_unistd.ml [hunk 3]
```diff
@@ -68,115 +68,125 @@ let lseek =
...
-  let c = local ~check_errno:true (unix_unistd_write_ptr ())
+  let c = local (unix_unistd_write_ptr ())
     PosixTypes.(fd @-> ptr void @-> size_t @-> returning size_t)
   in
   fun fd buf count ->
-    try
-      Size_t.to_int (c fd buf (Size_t.of_int count))
-    with Unix.Unix_error(e,_,_) -> raise (Unix.Unix_error (e,"write",""))
+    Errno_unix.raise_on_errno ~call:"write" begin fun () ->
+    Size_t.to_int (c fd buf (Size_t.of_int count))
```
Comment (dsheets) [inline lib/ctypes/unix_unistd.ml]: and here

**Comment by dsheets** [comment]
This is nice. Thank you. Could you please fix the indentation?

**Comment by dsheets** [comment]
(there are more places in the diff that I didn't mark because I got le tired)

**Comment by yallop** [comment]
I've fixed the indentation (3ad0961).

File: opam [hunk 4]
```diff
@@ -20,6 +20,7 @@ depopts: [
   "ctypes-foreign"
   "lwt"
   "unix-type-representations"
+  "unix-errno"
```
Comment (dsheets) [inline opam]: Does this need to have `base-unix`, too?

**Pull request merged**